File: 2020/src/main/java/com/mclaughlan/advent/days/Day2.java
package com.mclaughlan.advent.days;

import com.mclaughlan.advent.days.day2.Rule;
import com.mclaughlan.advent.days.day2.RulePositions;
import com.mclaughlan.advent.solution.Solution;

import java.util.stream.Collectors;

public class Day2 extends Solution {
    public Day2() {
        super("day2/input");
    }

    @Override
    public Object part1() {
        var lines = inputReader.streamLines();
        return lines.filter(this::checkPassword).collect(Collectors.toList()).size();
    }

    public Object part2() {
        var lines = inputReader.streamLines();
        return lines.filter(this::checkPassword2).collect(Collectors.toList()).size();
    }

    private boolean checkPassword(String line) {
        var parts = line.split(":");
        var rule = new Rule(parts[0]);
        return rule.isValid(parts[1]);
    }

    private boolean checkPassword2(String line) {
        var parts = line.split(":");
        var rule = new RulePositions(parts[0]);
        return rule.isValid(parts[1]);
    }
}

File: 2021/day2.py
from util.io import get_lines


def part1():
    h = 0
    d = 0

    for line in get_lines(2):
        parts = line.split(' ')
        cmd = parts[0]
        amount = int(parts[1])

        if cmd == 'forward':
            h += amount
        elif cmd == 'down':
            d += amount
        elif cmd == 'up':
            d -= amount

    return h * d


def part2():
    h = 0
    d = 0
    a = 0

    for line in get_lines(2):
        parts = line.split(' ')
        cmd = parts[0]
        amount = int(parts[1])

        if cmd == 'forward':
            h += amount
            d += a * amount
        elif cmd == 'down':
            a += amount
        elif cmd == 'up':
            a -= amount

    return h * d


print(f'Part 1: {part1()}')
print(f'Part 2: {part2()}')
